tempesta_fw/http_parser.c

```diff
@@ -2428,274 +2428,237 @@ __req_parse_referer(TfwHttpMsg *hm, unsigned char *data, size_t len)
-#define SB_APR		(SB_MAR + 31 * SEC24H)
-#define SB_MAY		(SB_APR + 30 * SEC24H)
-#define SB_JUN		(SB_MAY + 31 * SEC24H)
-#define SB_JUL		(SB_JUN + 30 * SEC24H)
-#define SB_AUG		(SB_JUL + 31 * SEC24H)
-#define SB_SEP		(SB_AUG + 31 * SEC24H)
-#define SB_OCT		(SB_SEP + 30 * SEC24H)
-#define SB_NOV		(SB_OCT + 31 * SEC24H)
-#define SB_DEC		(SB_NOV + 30 * SEC24H)
-/* Number of days before epoch including leap years. */
-#define EPOCH_DAYS	(1970 * 365 + 1970 / 4 - 1970 / 100 + 1970 / 400)
-
-static int
-__year_day_secs(unsigned int year, unsigned int day_sec)
+static time_t
+__year_day_secs(unsigned int year, unsigned int month, unsigned int day,
+                unsigned int hour, unsigned int min, unsigned int sec)
 {
-	unsigned int days = year * 365 + year / 4 - year / 100 + year / 400;
-
-	/* Add SEC24H if the year is leap and we left Feb behind. */
-	if (year % 4 == 0 && !(year % 100 == 0 && year % 400 != 0))
-		day_sec += SEC24H;
-
-	if (days < EPOCH_DAYS)
-		return -1;
-
-	return (days - EPOCH_DAYS) * SEC24H + day_sec;
+	time_t days;
+
+	year -= month <= 2;
+	/* Days in the current year since March 1 */
+	days = (153 * (month + (month > 2 ? -3 : 9)) + 2) / 5 + day - 1;
+	/* Days from March 1, 1 BC till March 1 of the current year */
+	days += year * 365 + year / 4 - year / 100 + year / 400;
+	/* 31 and 28 days were in January and February 1970 */
+	return (days - EPOCH_DAYS + 31 + 28) * SEC24H +
+	       hour * 3600 + min * 60 + sec;
 }
 
-static size_t
-__skip_weekday(unsigned char *p, size_t len)
+static time_t
+__parse_date(char *date_str, TfwHttpDateFmt *fmt)
 {
-	unsigned char lc, *c, *end = p + len;
+	static unsigned mday[] = { 31, 28, 31, 30, 31, 30,
+	                           31, 31, 30, 31, 30, 31 };
+	time_t year = 0, month = 0, day = 0, hour = 0, min = 0, sec = 0;
+	unsigned int month_int;
+
+	if (likely(fmt->type != RFC_850)) {
+		year = (date_str[fmt->year_pos + 0] - '0') * 1000 +
+		       (date_str[fmt->year_pos + 1] - '0') * 100 +
+		       (date_str[fmt->year_pos + 2] - '0') * 10 +
+		       (date_str[fmt->year_pos + 3] - '0');
+	} else {
+		year = (date_str[fmt->year_pos + 0] - '0') * 10 +
+		       (date_str[fmt->year_pos + 1] - '0');
+		year += (year < 70) ? 2000 : 1900;
+	}
+
+	if (likely(fmt->type != ISOC || date_str[fmt->day_pos] != ' '))
+	{
+		day = (date_str[fmt->day_pos + 0] - '0') * 10 +
+		      (date_str[fmt->day_pos + 1] - '0');
+	} else {
+		day = date_str[fmt->day_pos + 1] - '0';
+	}
+	hour = (date_str[fmt->time_pos + 0] - '0') * 10 +
+	       (date_str[fmt->time_pos + 1] - '0');
+	min  = (date_str[fmt->time_pos + 3] - '0') * 10 +
+	       (date_str[fmt->time_pos + 4] - '0');
+	sec  = (date_str[fmt->time_pos + 6] - '0') * 10 +
+	       (date_str[fmt->time_pos + 7] - '0');
+
+	month_int = TFW_CHAR4_INT(' ',
+	                          date_str[fmt->month_pos + 0],
+	                          date_str[fmt->month_pos + 1],
+	                          date_str[fmt->month_pos + 2]);
```

Comment:
I thinks compiler is smart enough to place `date_str[fmt->time_pos]` in a register, but I don't see any reason for the such massive code copy & paste - why not to use something like (please chose more appropriate names):
```C
unsigned char *p = date_str + fmt->year_pos;
year = (time_t)p[0] * 1000 + (p[1] - '0') * 100 + (p[2] - '0') * 10 + (p[2] - '0');
```
-- much less typing, isn't it?